sound/soc/sof/sof-audio.c

```diff
@@ -342,27 +349,31 @@ sof_prepare_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget
  * (DAI type for capture, AIF type for playback)
  */
 static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dapm_widget *widget,
-				    int dir)
+				    int dir, struct snd_soc_dapm_widget_list *list)
 {
 	struct snd_soc_dapm_path *p;
 	int err;
 	int ret = 0;
 
+	if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
+		err = sof_widget_free(sdev, widget->dobj.private);
+		if (err < 0)
+			ret = err;
+	}
+
 	/* free all widgets even in case of error to keep use counts balanced */
 	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
 		if (!p->walking && p->sink->dobj.private && widget->dobj.private) {
+			if (!widget_in_list(list, p->sink))
+				continue;
+
 			p->walking = true;
-			if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
-				err = sof_widget_free(sdev, widget->dobj.private);
-				if (err < 0)
-					ret = err;
-			}
```

Comment:
for my education, why did you move this block to before the loop @ranj063 ? It's not clear to me why we special case

Reply:
@plbossart this was a mistake in the first place. Assuming we have a DAI copier connected to 2 different host copiers, we would set up the DAI copier widget twice, once for each sink path.

Reply:
should this be treated as a separate fix then, and in a second patch add the check on the widget path? It's not clear to be if the two parts are strictly related?

Reply:
yes, fixed now. Thanks!

Reply:
ok, thanks